src/xss-common-keyvaluesort.hpp

```diff
@@ -16,6 +16,63 @@
 #include <omp.h>
 #endif
 
+/*
+ * Sort all the NAN's to end of the array and return the index of the last elem
+ * in the array which is not a nan
+ */
+template <typename T1, typename T2, typename vtype>
+X86_SIMD_SORT_INLINE arrsize_t move_nans_to_end_of_array(T1 *keys,
+                                                         T2 *vals,
+                                                         arrsize_t size)
+{
+    using reg_t = typename vtype::reg_t;
+
+    arrsize_t jj = size - 1;
+    arrsize_t ii = 0;
+    arrsize_t count = 0;
+
+    while (ii + vtype::numlanes < jj) {
+        reg_t in = vtype::loadu(keys + ii);
+        auto nanmask = vtype::convert_mask_to_int(
+                vtype::template fpclass<0x01 | 0x80>(in));
+
+        // Check if there are any nans in this vector, and process them if so
+        if (nanmask != 0x00) {
+            for (size_t offset = 0; offset < vtype::numlanes; offset++) {
+                if (is_a_nan(keys[ii])) {
+                    std::swap(keys[ii], keys[jj]);
+                    std::swap(vals[ii], vals[jj]);
+                    jj -= 1;
+                    count++;
+                }
+                else {
+                    ii += 1;
+                }
+            }
+        }
+        else {
+            ii += vtype::numlanes;
+        }
+    }
+
+    // Handle the remainders once we have less than 1 vector worth
+    while (ii < jj) {
```

Comment:
Could have combined this in the main loop with a masked load, but this is fine too.